CenterForOpenScience/ember-osf-web
# [ENG-3032] Added UpdateDropdown component for Outcome Reporting

-   Ticket: [ENG-3032]
-   Feature flag: n/a

Purpose:

The purpose of this component is to display revisions following the Versioning Implementation Plan found here: https://www.notion.so/Versioning-implementation-plan-6601412d86a34af4adbdb369b3482f0d#6576679d6859423985c74ef6f5f17181 on the front end UI for the NSF Eager Outcome Reporting project.  

Summary of Changes:

-app/models/revision.ts
	+Added versionNumber field to the revision model 
-lib/osf-components/addon/components/registries/update-dropdown/component.ts
        +created the update dropdown component class, added a task to retrieve a list of all revisions associated with a Registration
-lib/osf-components/addon/components/registries/update-dropdown/template.hbs
        +created a responsive dropdown handlebars file with a trigger button with UI display caret to display the list of revisions
-lib/osf-components/app/components/registries/update-dropdown/component.js:
        +exported the component file for update dropdown
-lib/osf-components/app/components/registries/update-dropdown/template.js
        +exported the handlebars file for update dropdown 
-lib/registries/addon/overview/-components/overview-topbar/template.hbs
	+Added the UpdateDropdown component to the OverviewTopbar handlebars file
+lib/registries/addon/overview/controller.ts
	+Added added the Intl service, the revision model, and the alias for revision model to the controller class 
+mirage/factories/revision.ts
	+Added the versionNumber and the dateCreated fields to the MirageRevisionModel
-mirage/scenarios/registrations.ts
	+Added versionNumber to the Mirage scenarios for demo purposes
-mirage/views/revision.ts
	+Added reviewState to the Mirage view for demo purposes
-translations/en-us.yml
	+Added a translation string for the dropdown

Screenshots:

![image](https://user-images.githubusercontent.com/82047646/128921677-44877f35-da8b-4596-ad4d-e918df353b4e.png)

![image](https://user-images.githubusercontent.com/82047646/128921726-0dc730a5-9aa0-4e41-a7d9-ec474a215925.png)

Side Effects:

The component will be placed directly to the right of the RegistriesStates component inside the parent OverviewTopbar component. This has the side affect of visually affecting the UI when the user clicks the caret in the form of a responsive dropdown list.

QA Notes:

-What pages should be tested?
	The Registries page should be tested to ensure that there are no discernible changes in the UI other than the width and height of the element; it should not exceed the OverviewTopbar in any way and in general look appropriate for the larger framework. It should also not overly obscure the original Registration page; pagination is currently being considered to handle the display of a large amount of edits.	
-Is cross-browser testing required/recommended?
	 Yes, always: Google Chrome, Microsoft Edge, Mozilla Firefox.
- What edge cases should QA be aware of?
	Edge cases currently include those for registrations with many edits and if there are no edits.
- What level of risk would you expect these changes to have?
	For security, with the http get request there is the potential for privilege escalation attacks where the user is able to gain access to sensitive data from a non-public repository or other non-privileged data about a Registration or project and its Contributor’s data provided to the OSF. Additionally, if there are any security misconfigurations in any of the environmental or config files those would also expose vulnerabilities. 

[ENG-3032]: https://openscience.atlassian.net/browse/ENG-3032
